cpu/esp32/periph/uart.c

```diff
@@ -279,63 +277,151 @@ void IRAM __uart_intr_handler (void *arg)
+    while (_uarts[uart].regs->status.txfifo_cnt >= UART_FIFO_MAX) {}
 
     /* send the byte by placing it in the TX FIFO using MPU */
     WRITE_PERI_REG(UART_FIFO_AHB_REG(uart), data);
 }
 
-static void __uart_intr_enable(uart_t uart)
+static void _uart_intr_enable(uart_t uart)
 {
-    __uarts[uart].regs->int_ena.rxfifo_full = 1;
-    __uarts[uart].regs->int_clr.rxfifo_full = 1;
-    __uarts[uart].used = true;
+    _uarts[uart].regs->int_ena.rxfifo_full = 1;
+    _uarts[uart].regs->int_clr.rxfifo_full = 1;
+    _uarts[uart].used = true;
 
-    DEBUG("%s %08x\n", __func__, __uarts[uart].regs->int_ena.val);
+    DEBUG("%s %08x\n", __func__, _uarts[uart].regs->int_ena.val);
 }
 
-/* systemwide UART initializations */
-void uart_system_init (void)
+void _uart_config (uart_t uart)
 {
-    for (unsigned uart = 0; uart < UART_NUMOF; uart++) {
-        /* reset all UART interrupt status registers */
-        __uarts[uart].regs->int_clr.val = ~0;
+    CHECK_PARAM (uart < UART_NUMOF);
+
+    /* setup the baudrate */
+    if (uart == UART_DEV(0) || uart == UART_DEV(1)) {
+        /* for UART0 and UART1, we can us the ROM function */
+        uart_div_modify(uart, (UART_CLK_FREQ << 4) / _uarts[uart].baudrate);
+    }
+    else if (_uart_set_baudrate(uart, _uarts[uart].baudrate) != UART_OK) {
+        return;
     }
-}
 
-void uart_print_config(void)
-{
-    for (unsigned uart = 0; uart < UART_NUMOF; uart++) {
-        ets_printf("\tUART_DEV(%d)\ttxd=%d rxd=%d\n", uart,
-                   __uarts[uart].pin_txd, __uarts[uart].pin_rxd);
+    /* set number of data bits, stop bits and parity mode */
+    if (_uart_set_mode(uart, _uarts[uart].data, _uarts[uart].stop,
+                             _uarts[uart].parity) != UART_OK) {
+        return;
+    }
+
+    /* reset the FIFOs */
+    _uarts[uart].regs->conf0.rxfifo_rst = 1;
+    _uarts[uart].regs->conf0.rxfifo_rst = 0;
+    _uarts[uart].regs->conf0.txfifo_rst = 1;
+    _uarts[uart].regs->conf0.txfifo_rst = 0;
+
+    if (_uarts[uart].isr_ctx.rx_cb) {
+        /* since reading can only be done byte by byte, we set
+           UART_RXFIFO_FULL_THRHD interrupt level to 1 byte */
+        _uarts[uart].regs->conf1.rxfifo_full_thrhd = 1;
+
+        /* enable the RX FIFO FULL interrupt */
+        _uart_intr_enable (uart);
+
+        /* route all UART interrupt sources to same the CPU interrupt */
+        intr_matrix_set(PRO_CPU_NUM, _uarts[uart].int_src, CPU_INUM_UART);
+
+        /* we have to enable therefore the CPU interrupt here */
+        xt_set_interrupt_handler(CPU_INUM_UART, _uart_intr_handler, NULL);
+        xt_ints_on(BIT(CPU_INUM_UART));
     }
 }
 
-int uart_set_baudrate(uart_t uart, uint32_t baudrate)
+int _uart_set_baudrate(uart_t uart, uint32_t baudrate)
 {
     DEBUG("%s uart=%d, rate=%d\n", __func__, uart, baudrate);
 
     CHECK_PARAM_RET (uart < UART_NUMOF, -1);
 
+    /* wait until TX FIFI is empty */
+    while (_uarts[uart].regs->status.txfifo_cnt != 0) { }
+
+    critical_enter();
```

Comment:
Can you put some information on why you added the critical*, does it apply to the *_mode or is it something separate

Reply:
It has to be ensured that the reconfiguration cannot be interrupted. Even though it is something I realized when I was testing `_uart_set_mode`, I think it has to be ensured for reconfiguring the baudrate too. That's why I added it with this PR.